prow/test/integration/test/webhook_server_test.go

```diff
@@ -118,8 +126,151 @@ func TestWebhookServerMutateProwjob(t *testing.T) {
+			Name: testServerAvailability,
+		}
+		if err := kubeClient.Get(ctx, key, &testProwjob); err != nil {
+			logrus.WithError(err).Infof("prowjob not found")
+		} else {
+			out, err := exec.Command("kubectl", "--context="+getClusterContext(), "delete", "prowjob", testServerAvailability).CombinedOutput()
+			if err != nil {
+				logrus.WithError(err).Infof("could not clean up cluster %s", string(out))
+			}	
+		}
+		
+	})
+	
+	key := types.NamespacedName{
+		Namespace: "default",
+		Name:      "webhook-server",
+	}
+	var found bool
+	var webhookDeployment appsv1.Deployment
+	err = kubeClient.Get(ctx, key, &webhookDeployment)
+	if err != nil {
+		t.Fatalf("could not list deployment %v", err)
+	}
+	found = true
+	originalProwJob := &v1.ProwJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ProwJob",
+			APIVersion: "prow.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				"foo":               "bar",
+				"default-me":        "enabled",
+				"admission-webhook": "enabled",
+			},
+			Name:      testServerAvailability,
+			Namespace: "default",
+		},
+		Spec: v1.ProwJobSpec{
+			Cluster: "foo",
+			Refs: &v1.Refs{
+				Repo: "bar",
+			},
+			Agent:  v1.TektonAgent,
+			Type:   v1.PresubmitJob,
+			Report: true,
+		},
+		Status: v1.ProwJobStatus{
+			State: v1.PendingState,
+		},
+	}
+	webhookDeploymentCopy := webhookDeployment.DeepCopy()
+	webhookDeploymentCopy.Spec.Template.Spec.Containers[0].LivenessProbe.PeriodSeconds = 4
+	var podList corev1.PodList
+	var webhookServerPod corev1.Pod
+	err = kubeClient.List(ctx, &podList, &client.ListOptions{})
+	if err != nil {
+		t.Fatalf("could not list pods %v", err)
+	}
+	for _, pod := range podList.Items {
+		if pod.ObjectMeta.Labels["app"] == "webhook-server" {
+			webhookServerPod = pod
+			found = true
+			break
+		}
+	}
+	var wg sync.WaitGroup
+	if !found {
+		t.Fatalf("webhook-server is not running")
+	} else {
+		wg.Add(1)
+		oldName := webhookServerPod.GetName()
+		go func() {
+			defer wg.Done()
+			for {
+				exist, newPod, err := checkNewPodInCluster(oldName, kubeClient, ctx)
+				if err != nil {
+					return
+				}
+				count := 0
+				if exist && newPod.Status.Phase == "Running" {
+					return
+				}
+				//update foo label of prowjob
+				originalProwJobCopy := originalProwJob.DeepCopy()
+				originalProwJobCopy.Labels["foo"] = strconv.Itoa(count)
+				err = kubeClient.Create(ctx, originalProwJobCopy, &client.CreateOptions{})
+				if err != nil {
+					t.Errorf("prowjob could not be created. Error: %v ", err)
+					return
+				}
+				out, err := exec.Command("kubectl", "--context="+getClusterContext(), "delete", "prowjob", testServerAvailability).CombinedOutput()
+				if err != nil {
+					t.Errorf("prowjob was not properly validated. Error: %v Output: %s", err, string(out))
+					return
+				}
+				count++
+			}
+		}()
+		if err := kubeClient.Patch(ctx, webhookDeploymentCopy, client.MergeFrom(&webhookDeployment), &client.PatchOptions{}); err != nil {
```

Comment:
I still don't understand why throwing heavy lifting process in a go routine while only keep this helper function running in main thread, is there something missing?

Reply:
we need to force a rolling update of the pod

Reply:
the same effect can be achieved by:
```
go func() {
  kubeClient.Patch(ctx, webhookDeploymentCopy, client.MergeFrom(&webhookDeployment), &client.PatchOptions{})
}

// Force rolling update of the pod
```

then you don't need to `t.Errorf` in go routine

Reply:
@cjwagner suggested this pattern

Reply:
> @cjwagner suggested this pattern

Sorry but this is not an answer which could convince me that this pattern is either technically or logically better than the approach I proposed.

Reply:
the idea is to force a rolling update of the pod while still continuously making calls to the webhook-server up until a new pod is available. I wasn't sure how to go about this in the right way so I asked for Cole's opinion and this is what he suggested. thats all I'm saying.

Reply:
thank you for clarifying! I'll be more specific on my suggestion:
```
go func() {
  time.Sleep(3 * time.Second) // Sleep to give for loop some time
  kubeClient.Patch(ctx, webhookDeploymentCopy, client.MergeFrom(&webhookDeployment), &client.PatchOptions{})
}()


for {
                        var isNewPod bool
			exist, newPod, err := checkNewPodInCluster(oldName, kubeClient, ctx)
			if err != nil {
				t.Fatalf("Failed checking for new pod: %v", err)
			}
			if exist && newPod.Status.Phase == "Running" {
				isNewPod = true
			}
			if err := kubeClient.Create(ctx, originalProwJob, &client.CreateOptions{}); err != nil {
				t.Fatalf("Prowjob could not be created. Error: %v ", err)
			}

                         if isNewPod {
                             return
                         }

			out, err := kubeClient.Delete(ctx, originalProwJob) // syntax might not be correct
			if err != nil {
				t.Fatalf("Could not delete prowjob: %v", err)
			}
		}
```